blog: Podman and Docker Rootless in DDEV, fixes #453 (#476)
ddev/ddev.com: src/content/blog/podman-and-docker-rootless.md

- **Modernized SSH authentication** ([#7511](https://github.com/ddev/ddev/pull/7511)): The `ddev auth ssh` command used to call `docker run` directly. We rewrote it to use the Docker API, making it compatible with alternative runtimes.
- **Optimized API call performance** ([#7587](https://github.com/ddev/ddev/pull/7587)): DDEV's Docker API logic was inefficient, making redundant calls without caching. We restructured the code to cache data and reduce unnecessary API requests.
- **Removed legacy docker-compose features** ([#7642](https://github.com/ddev/ddev/pull/7642)): Podman refuses to work with deprecated `links` and `external_links` directives in `docker-compose` files. We removed these legacy features and modernized DDEV's compose file generation.
- **Added Podman and Docker Rootless support** ([#7702](https://github.com/ddev/ddev/pull/7702)): DDEV now detects and supports Podman (rootful and rootless) and Docker Rootless. We added handling for Podman-specific limitations and enabled rootless environments to work without root privileges.

> These changes enabled Podman and Docker Rootless support.

Reviewer: An amazing journey, congratulations!

## Understanding Docker and Podman

### Why Choose Rootless?

Running containers without root privileges is more secure. Traditional Docker and rootful Podman need elevated privileges, which creates security risks in corporate environments where strict security policies apply.

Rootless alternatives (Podman Rootless and Docker Rootless) run containers without root access. This means:

- No root daemon running on your system
- Container processes cannot access root-owned files
- Reduced attack surface if a container is compromised

> Podman is rootless by default. Docker Rootless requires special setup.

Reviewer: It is probably worth explaining here that DDEV's traditional support has had its containers run as an unprivileged user, limiting the attack space. DDEV's built-in containers do not use root privileges (although docker itself does need root by default)

### The Socket Requirement for DDEV

DDEV needs a running Podman socket. The socket lets DDEV use the Docker API to talk to Podman, so DDEV can support both Docker and Podman with the same code.

## Installing Podman

### Podman Installation Options

> Podman can be installed with or without a GUI:

Reviewer: I think we probably need a little more than this. I missed both `podman init` and `podman start`. Maybe one of those collapsed sections with basic instructions on macOS and Linux? The majority of our users are macOS, but probably the majority of interested users are on Linux.

Author: I'll check the formatting options for Markdown here.

**Without GUI**:

- [macOS](https://podman.io/docs/installation#macos)
- [Windows](https://podman.io/docs/installation#windows)
- [Linux](https://podman.io/docs/installation#installing-on-linux)

**With Podman Desktop GUI**:

> - [macOS](https://podman-desktop.io/docs/installation/macos-install)

Reviewer: Interesting that they discourage use of homebrew.

- [Windows](https://podman-desktop.io/docs/installation/windows-install)
- [Linux](https://podman-desktop.io/docs/installation/linux-install)

For more information, see the [Podman tutorials](https://github.com/containers/podman/tree/main/docs/tutorials#readme).

### Installing Docker CLI

DDEV requires the `docker` CLI client for `ddev utility test` (this may change in the future). If you're using Podman without Docker, install only the CLI:

**macOS**:

```bash
brew install docker
```

**Linux (Ubuntu/Debian example)**:

1. [Set up Docker's apt repository](https://docs.docker.com/engine/install/ubuntu/#install-using-the-repository)
2. Install only the CLI:

   ```bash
   sudo apt-get install docker-ce-cli
   ```

   Note: You don't need to install `docker-ce` (the Docker engine).

## Configuring Podman for DDEV

After installing Podman, enable the API socket so DDEV can connect.

> ### Podman Rootless on Linux

Reviewer: Should this say "Linux and WSL2" ?

1. Create the `podman` group:

   ```bash
   sudo groupadd podman
   ```

2. Add your user to the group:

   ```bash
   sudo usermod -aG podman $USER
   ```

3. Log out and log back in, or activate the changes immediately:

   ```bash
   newgrp podman
   ```

   If you're in a VM, restart it for changes to take effect.

4. Configure Podman socket permissions:

   ```bash
   # Configure permissions for /run/podman
   sudo tee /etc/tmpfiles.d/podman.conf > /dev/null <<EOF
   d /run/podman 0770 root podman
   EOF

   # Configure socket to run with podman group
   sudo mkdir -p /etc/systemd/system/podman.socket.d
   sudo tee /etc/systemd/system/podman.socket.d/override.conf > /dev/null <<EOF
   [Socket]
   SocketMode=0660
   SocketUser=root
   SocketGroup=podman
   EOF

   # Apply changes
   sudo systemd-tmpfiles --create
   sudo systemctl daemon-reload
   ```

5. Enable and start the socket:

   ```bash
   sudo systemctl enable --now podman.socket
   ```

6. Verify the socket:

   ```bash
   ls /var/run/podman/podman.sock
   ```

   Or check with:

   ```bash
   sudo podman info --format '{{.Host.RemoteSocket.Path}}'
   ```

> 7. Configure Docker to use Podman. Choose one of these methods:

Reviewer: This statement will be confusing. I assume this means "Configure `docker` CLI to connect to Podman instead of Docker Daemon."

   **Option A: Create Docker context** (if you have `docker` CLI installed):

   ```bash
   # View existing contexts
   docker context ls

   # Create Podman rootful context
   docker context create podman \
       --description "Podman (rootful)" \
       --docker host="unix:///var/run/podman/podman.sock"

   # Switch to the new context
   docker context use podman

   # Verify it works
   docker ps
   ```

   **Option B: Set environment variable in your shell profile** (`~/.bashrc` or `~/.zshrc`):

   ```bash

>    export DOCKER_HOST=unix:///var/run/podman/podman.sock

Reviewer: I wonder if we should not mention this technique. It seems like it will be fragile and little used.

   ```bash
   systemctl --user enable --now docker.socket
   ```

5. Verify the socket:

   ```bash
   ls $XDG_RUNTIME_DIR/docker.sock
   ```

   You should see `/run/user/1000/docker.sock` (the number may vary).

6. Configure Docker context (if needed):

   ```bash
   # View existing contexts
   docker context ls

   # Create rootless context
   docker context create rootless \
       --description "Rootless runtime socket" \
       --docker host="unix://$XDG_RUNTIME_DIR/docker.sock"

   # Switch to the context
   docker context use rootless

   # Verify it works
   docker ps
   ```

7. Docker Rootless requires no-bind-mounts mode

   Docker Rootless has a limitation with bind mounts that affects DDEV. You must enable `no-bind-mounts` mode:

   ```bash
   ddev config global --no-bind-mounts=true
   ```

   **Why this is needed**:

   Docker Rootless sets ownership for bind mounts to `root` inside containers. This is a known issue:
   - [Mounting a volume with rootless always assigns ownership to root](https://github.com/moby/moby/issues/45919)
   - [Add ability to mount volume as user other than root](https://github.com/moby/moby/issues/2259)

>    The `root` user inside the container maps to your host user, but many services will not run as root:

Reviewer: It's interesting that this is the same classic problem we've had with Docker Desktop for Linux and for a time with virtiofs.

For example, here's a system with four active Docker contexts:

```bash
$ docker context ls
NAME                DESCRIPTION                               DOCKER ENDPOINT
default             Current DOCKER_HOST based configuration   unix:///var/run/docker.sock
podman              Podman (rootful)                          unix:///var/run/podman/podman.sock
podman-rootless *   Podman (rootless)                         unix:///run/user/1000/podman/podman.sock
rootless            Rootless runtime socket                   unix:///run/user/1000/docker.sock
```

Switch between them with:

```bash
docker context use <context-name>
```

**Note**: Running both Docker and Podman in rootful mode at the same time may cause network conflicts. See [Podman and Docker network problem on Fedora 41](https://github.com/containers/podman/issues/24486).

## Switching Runtimes with DDEV

DDEV automatically detects your active container runtime. To switch:

1. Stop DDEV projects:

   ```bash
   ddev poweroff
   ```

2. Switch Docker context or change the `DOCKER_HOST` environment variable

3. Start your project:

   ```bash
   ddev start
   ```

> ## Which Runtime Should You Choose?

Reviewer: Great section. However, it should encourage standard Docker setup first, with link to the docs, as it's by far the best for local development.

   docker ps
   ```

   For more details, see [Podman rootless tutorial](https://github.com/containers/podman/blob/main/docs/tutorials/rootless_tutorial.md)

4. Proceed with [DDEV installation](https://docs.ddev.com/en/stable/users/install/ddev-installation/#ddev-installation-linux).

### Podman Rootless Performance Optimization

Podman Rootless is significantly slower than Docker. See these resources:

- [Podman run/build performance issues](https://github.com/containers/podman/issues/13226)
- [Podman Performance documentation](https://github.com/containers/podman/blob/main/docs/tutorials/performance.md)

To improve performance, install `fuse-overlayfs` and configure the overlay storage driver:

**Install fuse-overlayfs**:

```bash
# Ubuntu/Debian
sudo apt-get install fuse-overlayfs
# Fedora
sudo dnf install fuse-overlayfs
```

**Configure storage**:

```bash
cat << 'EOF' > ~/.config/containers/storage.conf
[storage]
driver = "overlay"
[storage.options.overlay]
mount_program = "/usr/bin/fuse-overlayfs"
EOF
```

**Warning**: If you already have Podman containers, images, or volumes, you'll need to reset Podman for this change to take effect:

```bash
podman system reset
```

> This removes all existing containers, images, and volumes (similar to `docker system prune -a`).

Author: I tested it on Arch Linux, and `docker ps` worked fine without reboot after `podman system reset`.

> but after reboot I could

Maybe this command can help to do it without reboot:

```bash
systemctl --user restart podman.socket
```

---
title: "Podman and Docker Rootless in DDEV"
pubDate: 2026-01-08
#modifiedDate: 2026-01-08
summary: After years of development, DDEV now supports Podman and Docker Rootless for secure, rootless container development.
author: Stas Zhuk
featureImage:
  src: /img/blog/2026/01/ddev-podman-docker-rootless.png
  alt: DDEV logo with Podman and Docker logos
categories:
  - Announcements
  - Guides
---

> **TL;DR**: DDEV now supports Podman and Docker Rootless in [DDEV HEAD](https://docs.ddev.com/en/stable/developers/building-contributing/#testing-latest-commits-on-head) (coming in v1.25.0). Jump to setup instructions: [Linux/WSL2](#key-aim-linux-and-wsl2-users) · [macOS](#macos) · [Windows](#windows)

Reviewer: Let's expand this.

```suggestion
**TL;DR**: DDEV now supports Podman and Docker Rootless in [DDEV HEAD](https://docs.ddev.com/en/stable/developers/building-contributing/#testing-latest-commits-on-head) (coming in v1.25.0).  Podman and Docker Rootless are a bit more trouble than the [recommended normal traditional docker providers](https://docs.ddev.com/en/stable/users/install/docker-installation/) and have some serious trade-offs. On macOS you can't use the nomal default ports 80 and 443. On Linux Docker Rootless you can't bind-mount directories, so the entire project has to mutagen-synced.

Jump to setup instructions: [Linux/WSL2](#key-aim-linux-and-wsl2-users) · [macOS](#macos) · [Windows](#windows)
```

   **Option 1: Create a Docker context** (recommended, more flexible):

   ```bash
   # Create Podman context (path to socket may vary)
   # Use the socket path from `podman machine start` output
   docker context create podman-rootless \
       --description "Podman (rootless)" \
       --docker host="unix:///var/folders/x3/r1wk89cd3_x0yb_21dgnj53m0000gn/T/podman/podman-machine-default-api.sock"

   # Switch to the new context
   docker context use podman

   # Verify it works
   docker ps
   ```

   This approach uses Docker contexts to switch between different container runtimes without modifying system sockets. This is more flexible if you want to use multiple Docker providers.

   **Option 2: Use the default Docker socket** (simpler, but less flexible):

   ```bash
   # Install podman-mac-helper
   # Use the command from `podman machine start` output
   sudo /opt/homebrew/Cellar/podman/5.7.1/bin/podman-mac-helper install
   podman machine stop
   podman machine start

   # Verify it works
   docker ps
   ```

3. Proceed with [DDEV installation](https://docs.ddev.com/en/stable/users/install/ddev-installation/).

> 4. Handle privileged ports (<1024):

Author: Okay.

> I do wonder why our automatic switching to ephemeral ports doesn't work in this situation.

Since the ephemeral ports don't check the actual bind, the 80/443 ports are available for use.

To improve performance, install `fuse-overlayfs` and configure the overlay storage driver:

**Install fuse-overlayfs**:

```bash
# Ubuntu/Debian
sudo apt-get update && sudo apt-get install fuse-overlayfs
# Fedora
sudo dnf install --refresh fuse-overlayfs
```

**Configure storage**:

```bash
mkdir -p ~/.config/containers
cat << 'EOF' > ~/.config/containers/storage.conf
[storage]
driver = "overlay"
[storage.options.overlay]
mount_program = "/usr/bin/fuse-overlayfs"
EOF
```

**Warning**: If you already have Podman containers, images, or volumes, you'll need to reset Podman for this change to take effect:

```bash
podman system reset
```

This removes all existing containers, images, and volumes (similar to `docker system prune -a`).

### Configuring Podman Rootful

Rootless Podman is recommended for most users. However, if you need rootful Podman, the setup differs from rootless in two key ways:

1. User group permissions: configure [group permissions](https://github.com/podman-desktop/podman-desktop/issues/2861#issuecomment-1596192247) to allow non-root users to access the rootful socket

2. Socket activation and paths:

   ```bash
   # Enable rootful socket (requires sudo)

>    sudo systemctl enable --now podman.socket

Reviewer: With this approach we also need 
```
export DOCKER_HOST=unix:///var/run/podman/podman.sock
```

or instructions to create a context

   # Socket path for rootful
   ls /var/run/podman/podman.sock
   ```

   Compare to rootless:

   ```bash
   # Rootless socket (no sudo)
   systemctl --user enable --now podman.socket

   # Socket path for rootless

>    ls $XDG_RUNTIME_DIR/podman/podman.sock

Reviewer: I think the "compoare to rootless" will just confuse people and they'll try to follow these instructions.

I don't think XDG_RUNTIME_DIR is there on Ubuntu btw

       echo 'net.ipv4.ip_unprivileged_port_start=0' | sudo tee -a /etc/sysctl.d/60-rootless.conf
       sudo sysctl --system
     fi
   fi
   # Allow loopback connections (needed for working Xdebug)
   # See https://github.com/moby/moby/issues/47684#issuecomment-2166149845
   mkdir -p ~/.config/systemd/user/docker.service.d
   cat << 'EOF' > ~/.config/systemd/user/docker.service.d/override.conf
   [Service]
   Environment="DOCKERD_ROOTLESS_ROOTLESSKIT_DISABLE_HOST_LOOPBACK=false"
   EOF
   ```

3. Enable the Docker socket, and verify it's running:

   ```bash
   systemctl --user enable --now docker.socket

   # You should see `/run/user/1000/docker.sock` (the number may vary):
   ls $XDG_RUNTIME_DIR/docker.sock
   ```

4. Configure Docker API to use Docker rootless:

   ```bash
   # View existing contexts
   docker context ls

   # Create rootless context if it doesn't exist
   docker context inspect rootless >/dev/null 2>&1 || \
       docker context create rootless \
           --description "Rootless runtime socket" \
           --docker host="unix://$XDG_RUNTIME_DIR/docker.sock"

   # Switch to the context
   docker context use rootless

> 

Author: Maybe, but running it doesn't break anything, because I check for the context before creating it.

I didn't have any context created on Arch-based Linux (because I didn't run the script; it's not intended for Arch).

For more information, see the [official Podman installation guide for macOS](https://podman.io/docs/installation#macos) and [Podman tutorials](https://github.com/containers/podman/tree/main/docs/tutorials#readme).

### Installing Docker CLI

```bash
brew install docker
```

### Configuring Podman

After installing Podman, initialize and start the Podman machine:

```bash
podman machine init
podman machine start
```

The Podman machine will automatically configure the socket. Verify it's running:

```bash
podman info --format '{{.Host.RemoteSocket.Path}}'
```

If you have Docker CLI installed, you can create a context for Podman:

```bash
# Create Podman context
docker context create podman-rootless \
    --description "Podman (rootless)" \
    --docker host="unix://$(podman info --format '{{.Host.RemoteSocket.Path}}')"

# Switch to the new context
docker context use podman-rootless

# Verify it works
docker ps
```

Otherwise, set the `DOCKER_HOST` environment variable in your shell profile (`~/.zshrc` or `~/.bashrc`):

```bash
export DOCKER_HOST=unix://$(podman info --format '{{.Host.RemoteSocket.Path}}')

> ```

Author: I didn't test macOS, hoping for possible improvements or tips from @rfay in this section.

## Understanding Docker and Podman

### Open Source Alternatives to Docker Desktop

A common misconception is that Podman is the only open-source alternative to Docker Desktop. This is not true. There are several fully open-source alternatives available on every platform:

- **[Docker Engine](https://docs.docker.com/engine/)** - The original open-source Docker, free to use
- **[Rancher Desktop](https://rancherdesktop.io/)** - Open source container management with choice of dockerd or containerd
- **[Lima](https://lima-vm.io/)** - Linux virtual machines
- **[Colima](https://github.com/abiosoft/colima)** - Container runtime with minimal setup (built on Lima)
- **[Podman Desktop](https://podman-desktop.io/)** - GUI for Podman with Docker compatibility

All of these work with DDEV. The main reason to choose Podman specifically is if your organization forbids Docker entirely or if you want rootless operation by default.

> ### Why Choose Rootless?

Reviewer: Great section. Maybe "Why choose Podman" right after it. Or before it.

> Running containers without root privileges is more secure. Traditional Docker and rootful Podman need elevated privileges, which creates security risks in corporate environments where strict security policies apply.

Reviewer: ```suggestion
Although DDEV's use of all Docker providers is quite secure, and we run containers as normal users with limited privileges, the rootless approaches to Docker and Podman actually run the *docker daemon* without root privileges, closing additional attack surface. Traditional Docker and rootful Podman daemons need elevated privileges, which creates security risks in corporate environments where strict security policies apply. (Note that DDEV is targeted at local development, where there are few risks of specialized attacks using this vector anyway.)
```

   ```

5. Enable and start the socket:

   ```bash
   sudo systemctl enable --now podman.socket
   ```

6. Verify the socket:

   ```bash
   ls /var/run/podman/podman.sock
   ```

   Or check with:

   ```bash
   sudo podman info --format '{{.Host.RemoteSocket.Path}}'
   ```

7. Configure Docker API to use Podman. Choose one of these methods:

   **Option A: Create Docker context** (if you have `docker` CLI installed):

   ```bash
   # View existing contexts
   docker context ls

   # Create Podman rootful context
   docker context create podman \
       --description "Podman (rootful)" \
       --docker host="unix:///var/run/podman/podman.sock"

   # Switch to the new context
   docker context use podman

   # Verify it works
   docker ps
   ```

   **Option B: Set environment variable in your shell profile** (`~/.bashrc` or `~/.zshrc`):

   ```bash
   export DOCKER_HOST=unix:///var/run/podman/podman.sock
   ```

### Common Podman Issues

- [Podman rootless tutorial](https://github.com/containers/podman/blob/main/docs/tutorials/rootless_tutorial.md)
- [Arch Wiki](https://wiki.archlinux.org/title/Podman)

### Podman Rootless Performance Optimization

> Podman Rootless is slower than Docker. See these resources:

Reviewer: ```suggestion
Podman Rootless is significantly slower than Docker. See these resources:
```
